panic during log append under certain circumstances

**Describe the bug**
When:
- a node receives an append-logs RPC request from a valid leader
- the "entries" field of the request includes some overlapping entries already on the follower but not all of the follower's entries
- none of the overlapping entries are mismatched

The follower will panic as it overruns the "entries" array

**Desired behavior**
The follower should discard logs past those in a valid append-logs entry (this may also be mitigated in elections by improving log-comparison, but since this only affects uncommitted entries it could still happen

**Additional info**

Stacktrace:

```
2020-06-08T14:54:42-04:00 INF Election failed got=1 needed=2 success=false
panic: runtime error: index out of range [4] with length 4

goroutine 67 [running]:
github.com/btmorr/leifdb/internal/node.reconcileLogs(0xc000103ac0, 0xc000231110, 0xc00069f2e0)
	/Users/toma/raft/leifdb/internal/node/node.go:713 +0x47e
github.com/btmorr/leifdb/internal/node.(*Node).HandleAppend(0xc000170000, 0xc000231110, 0x1b)
	/Users/toma/raft/leifdb/internal/node/node.go:787 +0x16b
github.com/btmorr/leifdb/internal/raftserver.(*server).AppendLogs(0xc00018c1a0, 0x1b12220, 0xc0004f9020, 0xc000231110, 0xc00018c1a0, 0xc0004f9020, 0xc000098ba0)
	/Users/toma/raft/leifdb/internal/raftserver/rpc.go:27 +0xb8
github.com/btmorr/leifdb/internal/raft._Raft_AppendLogs_Handler(0x18b22e0, 0xc00018c1a0, 0x1b12220, 0xc0004f9020, 0xc000012060, 0x0, 0x1b12220, 0xc0004f9020, 0xc0001aa1b0, 0x82)
	/Users/toma/raft/leifdb/internal/raft/raft.pb.go:821 +0x217
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000216820, 0x1b1b8c0, 0xc000196780, 0xc00026a700, 0xc0004f8240, 0x31e2738, 0x0, 0x0, 0x0)
	/Users/toma/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1082 +0x50a
google.golang.org/grpc.(*Server).handleStream(0xc000216820, 0x1b1b8c0, 0xc000196780, 0xc00026a700, 0x0)
	/Users/toma/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1405 +0xccd
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0004b6860, 0xc000216820, 0x1b1b8c0, 0xc000196780, 0xc00026a700)
	/Users/toma/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/toma/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
```

--- a/internal/node/node.go
+++ b/internal/node/node.go
@@ -706,6 +706,10 @@ func reconcileLogs(logStore *raft.LogStore, body *raft.AppendRequest) *raft.LogS
 	if body.PrevLogIndex < int64(len(logStore.Entries)-1) {
 		overlappingEntries := logStore.Entries[body.PrevLogIndex+1:]
 		for i, rec := range overlappingEntries {
+			if i >= len(body.Entries) {
+				mismatchIdx = body.PrevLogIndex + int64(i)
+				break
+			}
 			if rec.Term != body.Entries[i].Term {
 				mismatchIdx = body.PrevLogIndex + 1 + int64(i)
 				break

--- a/internal/node/node_test.go
+++ b/internal/node/node_test.go
@@ -165,7 +165,7 @@ func TestReconcileLogs(t *testing.T) {
 
 	testCases := []ReconcileTestCase{
 		{
-			Name:  "Empty mind, empty body",
+			Name:  "Empty log and request",
 			Store: emptyLog,
 			Request: &raft.AppendRequest{
 				Term:         0,
@@ -176,7 +176,7 @@ func TestReconcileLogs(t *testing.T) {
 				Entries:      []*raft.LogRecord{}},
 			Expected: emptyLog},
 		{
-			Name:  "Empty mind, full body",
+			Name:  "Empty log, populated request",
 			Store: emptyLog,
 			Request: &raft.AppendRequest{
 				Term:         3,
@@ -187,7 +187,7 @@ func TestReconcileLogs(t *testing.T) {
 				Entries:      firstThree},
 			Expected: starterLog},
 		{
-			Name:  "Full mind, full body",
+			Name:  "Populated log and request",
 			Store: starterLog,
 			Request: &raft.AppendRequest{
 				Term:         6,
@@ -198,7 +198,18 @@ func TestReconcileLogs(t *testing.T) {
 				Entries:      nextTwo},
 			Expected: appendLog},
 		{
-			Name:  "Incongruous mind",
+			Name:  "Match but truncate",
+			Store: appendLog,
+			Request: &raft.AppendRequest{
+				Term:         6,
+				LeaderId:     "localhost:8181",
+				PrevLogIndex: 2,
+				PrevLogTerm:  3,
+				LeaderCommit: -1,
+				Entries:      []*raft.LogRecord{nextTwo[0]}},
+			Expected: &raft.LogStore{Entries: appendLog.Entries[:4]}},
+		{
+			Name:  "Mismatch and add",
 			Store: starterLog,
 			Request: &raft.AppendRequest{
 				Term:         6,
@@ -210,6 +221,11 @@ func TestReconcileLogs(t *testing.T) {
 			Expected: overlapLog}}
 
 	for _, tc := range testCases {
+		defer func() {
+			if r := recover(); r != nil {
+				t.Errorf("Recovered panic in %s: %v", tc.Name, r)
+			}
+		}()
 		result := reconcileLogs(tc.Store, tc.Request)
 		testutil.CompareLogs(t, tc.Name, result, tc.Expected)
 	}

